packages/flutter/lib/src/rendering/debug_overflow_indicator.dart

```diff
@@ -0,0 +1,301 @@
+      'This is considered an error condition because it indicates that there '
+      'is content that cannot be seen. If the content is legitimately bigger '
+      'than the available space, consider clipping it with a ClipRect widget '
+      'before putting it in the $runtimeType, or using a scrollable '
+      'container, like a ListView.';
+
+    final List<String> overflows = <String>[];
+    if (overflow.left > 0.0)
+      overflows.add('${_formatPixels(overflow.left)} pixels on the left');
+    if (overflow.top > 0.0)
+      overflows.add('${_formatPixels(overflow.top)} pixels on the top');
+    if (overflow.bottom > 0.0)
+      overflows.add('${_formatPixels(overflow.bottom)} pixels on the bottom');
+    if (overflow.right > 0.0)
+      overflows.add('${_formatPixels(overflow.right)} pixels on the right');
+    String overflowText = '';
+    assert(overflows.isNotEmpty,
+        "Somehow $runtimeType didn't actually overflow like it thought it did.");
+    switch (overflows.length) {
+      case 1:
+        overflowText = overflows.first;
+        break;
+      case 2:
+        overflowText = '${overflows.first} and ${overflows.last}';
+        break;
+      default:
+        overflows[overflows.length - 1] = 'and ${overflows[overflows.length - 1]}';
+        overflowText = overflows.join(', ');
+    }
+    FlutterError.reportError(
+      new FlutterErrorDetailsForRendering(
+        exception: 'A $runtimeType overflowed by $overflowText.',
+        library: 'rendering library',
+        context: 'during layout',
+        renderObject: this,
+        informationCollector: (StringBuffer information) {
+          information.writeln(overflowHints);
+          information.writeln('The specific $runtimeType in question is:');
+          information.writeln('  ${toStringShallow(joiner: '\n  ')}');
+          information.writeln('◢◤' * (FlutterError.wrapWidth ~/ 2));
+        },
+      ),
+    );
+  }
+
+  /// To be called when the overflow indicators should be painted.
+  ///
+  /// Typically only called if there is an overflow, and only from within a
+  /// debug build.
+  ///
+  /// See example code in [DebugOverflowIndicatorMixin] documentation.
+  void paintOverflowIndicator(
+      PaintingContext context, Offset offset, Rect containerRect, Rect childRect,
+      {String overflowHints}) {
+    final RelativeRect overflow = new RelativeRect.fromRect(containerRect, childRect);
+
+    if (overflow.left <= 0.0 &&
+        overflow.right <= 0.0 &&
+        overflow.top <= 0.0 &&
+        overflow.bottom <= 0.0) {
+      return;
+    }
+
+    final List<_OverflowRegionData> overflowRegions = _calculateOverflowRegions(overflow, containerRect);
+    for (_OverflowRegionData region in overflowRegions) {
+      context.canvas.drawRect(region.rect.shift(offset), _indicatorPaint);
+
+      _indicatorLabel.text = new TextSpan(
+        text: region.label,
+        style: _indicatorTextStyle,
+      );
+      _indicatorLabel.layout();
```

Comment:
having to call this every frame is expensive. It probably isn't critical for this since we don't expect these overflow labels to appear a lot even in debug mode, but if this was something that would render in release builds i'd recommend caching the label for each side separately so that you only have to call layout() on them once per layout (not multiple times per paint).